File: AOC_CS/Days/Day09.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace AOC_CS.Days
{
    class Day09
    {
        const string INPUT_FILE = "Inputs/day09.txt";

        const int PREAMBLE = 25;

        private long[] inputs;
        private HashSet<long>[] sums;

        public Day09()
        {
            inputs = null;
            sums = null;
        }

        private void ReadInput()
        {
            if (inputs != null && sums != null) return;

            var lines = File.ReadAllLines(INPUT_FILE);
            inputs = Array.ConvertAll(lines, long.Parse);

            sums = new HashSet<long>[inputs.Length];
            for (var i = 0; i < inputs.Length; ++i)
                sums[i] = new HashSet<long>();
            for (var i = 0; i < inputs.Length; ++i)
            {
                for (var j = 1; j < PREAMBLE && i + j < inputs.Length; ++j)
                {
                    var sum = inputs[i] + inputs[i + j];
                    for (var k = j + 1; k <= PREAMBLE && i + k < inputs.Length; ++k)
                        sums[i + k].Add(sum);
                }
            }
        }

        public long Solve1()
        {
            ReadInput();
            for (var i = PREAMBLE; i < inputs.Length; ++i)
            {
                if (!sums[i].Contains(inputs[i]))
                    return inputs[i];
            }
            return inputs.Length;
        }

        public long Solve2()
        {
            ReadInput();
            var num = Solve1();
            for (var i = 0; i < inputs.Length; ++i)
            {
                long sum = inputs[i], min = sum, max = sum;
                for (var j = i + 1; j < inputs.Length; ++j)
                {
                    sum += inputs[j];
                    if (sum > num) break;
                    if (inputs[j] < min) min = inputs[j];
                    if (inputs[j] > max) max = inputs[j];
                    if (sum == num) return min + max;
                }
            }
            return 0;
        }
    }
}

File: AOC_CPP/day09.cpp
#include <fstream>
#include <iostream>
#include <set>
#include <stdexcept>
#include <vector>

namespace AOC_CPP::Day09
{
    constexpr auto INPUT_FILE = "../Inputs/day09.txt";

    using IntType = int64_t;

    constexpr size_t PREAMBLE = 25;
    std::vector<IntType> inputs;

    std::vector<std::set<IntType>> sums;

    void read_input()
    {
        using namespace std::string_literals;
        inputs.clear();
        IntType n;
        std::ifstream infile(INPUT_FILE);
        if (!infile) throw std::runtime_error("File \""s + INPUT_FILE + "\" not found");
        while (infile >> n)
            inputs.push_back(n);
        infile.close();
        sums.assign(inputs.size(), std::set<IntType>{});
        for (size_t i = 0; i < inputs.size(); ++i)
        {
            for (size_t j = 1; j < PREAMBLE && i + j < inputs.size(); ++j)
            {
                const auto tmp = inputs[i] + inputs[i + j];
                for (size_t k = j + 1; k <= PREAMBLE && i + k < inputs.size(); ++k)
                    sums[i + k].insert(tmp);
            }
        }
    }

    IntType solve1()
    {
        for (size_t i = PREAMBLE; i < inputs.size(); ++i)
        {
            if (sums[i].find(inputs[i]) == sums[i].end())
                return inputs[i];
        }
        return 0;
    }

    IntType solve2()
    {
        const auto num = solve1();
        for (size_t i = 0; i < inputs.size(); ++i)
        {
            auto sum = inputs[i], min = sum, max = sum;
            for (size_t j = i + 1; j < inputs.size(); ++j)
            {
                const auto tmp = inputs[j];
                sum += tmp;
                if (sum > num) break;
                if (tmp > max) max = tmp;
                if (tmp < min) min = tmp;
                if (sum == num) return min + max;
            }
        }
        return 0;
    }
} // namespace AOC_CPP::Day09

int main()
{
    using namespace AOC_CPP;
    Day09::read_input();
    std::cout << Day09::solve1() << '\n';
    std::cout << Day09::solve2() << '\n';

    return 0;
}
